File: Assignment6/IntStack.java
import java.util.Arrays;

class IntStack {

  private int stack[];

  private int top;

  static int SIZE = 16; 

  static int MINSIZE = 1 << 15; 

  IntStack() {
    this(SIZE);
  }

  IntStack(final int size) {
    stack = new int[size];
    clear();
  }

  final int getTop() {
    return top;
  }

  final int setTop(final int top) {
    return this.top = top;
  }

  final void clear() {
 
    top = -1;
  }

  final boolean isEmpty() {
    return top < 0;
  }

  final void push(final int i) {

    if (++top >= stack.length) {
      expand();
    }
    stack[top] = i;
  }

  final int pop() {
    final int r = stack[top--];
    shrink();
    return r;
  }

  final int get(final int i) {
    return stack[i];
  }

  final void set(final int i, final int val) {
    stack[i] = val;
  }

  final int size() {
    return top + 1;
  }


  private final void expand() {
    final int l = stack.length;
    final int[] newstack = new int[l << 1];

    System.arraycopy(stack, 0, newstack, 0, l);
    stack = newstack;
  }


  private final void shrink() {
    int l = stack.length;
    if (l <= MINSIZE || top << 2 >= l)
      return;
    l = 1 + (top << 1); 
    if (top < MINSIZE) {
      l = MINSIZE;
    }

    final int[] newstack = new int[l];
    System.arraycopy(stack, 0, newstack, 0, top + 1);
    stack = newstack;
  }

  int[] toArray() {
    final int[] array = new int[size()];
    if (size() > 0) {
      System.arraycopy(stack, 0, array, 0, size());
    }
    return array;
  }

  public final void reverse() {
    int l = size();
    int h = l >> 1;
    for (int i = 0; i < h; i++) {
      int temp = stack[i];
      stack[i] = stack[l - i - 1];
      stack[l - i - 1] = temp;
    }
  }
 public static int[] randomm()  {
      int seed=88;int[] data;int random;
      data=new int[10];
    /*  for (int i=100,j=1;i<105;i++,j++){
          random=i*seed+seed*(i-seed);
          data[j]=random;
          System.out.println(random);
      }
      */
      
      for (int i=100,j=0;j<5;i=i*seed,j++){
          random=i*seed+seed*(i-seed);
          data[j]=random;
          System.out.println(random);
      }
      
      return data;
      }

  @Override
  public String toString() {
    return Arrays.toString(toArray());
  }

public static void main(final String[] args) {
    IntStack s=new IntStack();
    int[] a=randomm();
s.push(a[0]);
s.push(a[1]);
s.push(a[2]);
s.push(a[3]);
s.push(a[4]);
//s.push(a[5]);
System.out.println("The size of stack is: ");
System.out.println(s.getTop()+1);


  }
}

File: Assignment6/IntStack.cpp
#include<vector>
#include<iostream>
using namespace std;
class IntStack{
public:
	vector<int> stack;
	int top;
 int SIZE=16;
	int minsize =1<<(SIZE-1);
IntStack(){
	IntStack(SIZE);
}
IntStack(int SIZE){
	stack.reserve(SIZE);
	clear();
}
int getTop() {
	return top;
}
int setTop(int top) {
        return this->top = top;
}
void clear() {
        top = -1;
}
bool isEmpty() {
        return top < 0;
}
void push(int i){
	if(++top >=stack.size())
		expand();
	stack[top]=i;
}
int pop(){
	int r=stack[top--];
	shrink();
	return r;
}
int get(int i){
	return stack[i];
}
void set(int i,int val){
	stack[i]=val;
}
int size(){
	return top+1;
}
void expand(){
	int l=stack.size();
	stack.resize(l*4);
}
void shrink(){
	int l=stack.size();
	if(l<=minsize || top<<2 >=1)
		return;
	l=1+(top<< 1);
	if(top <minsize)
		l=minsize;
	stack.resize(l);
}
vector<int> toArray(){
	vector<int> array;
	if(size() > 0)
		array=stack;
	return array;
}
void reverse(){

	int l=size();
	int h=l>>1;
	for(int i=0;i<h;i++){
		int temp=stack[i];
		stack[i]=stack[l-i-1];
		stack[l-i-1]=temp;
	}
}
};
int * randomsdq(int seed){


        static int data[5];int random;

        for (int i=100,j=0;j<5;i=i*seed,j++){
            random=i*seed+seed*(i-seed);
            data[j]=random;
        }

        return data;

}

int main()
{
int* str1;int seed;
cout<<"enter seed";
cin>>seed;
str1=randomsdq(seed);
IntStack s(str1[0]);
s.push(str1[0]);
s.push(str1[1]);
s.push(str1[2]);
s.push(str1[3]);
s.push(str1[4]);
cout<<str1[0]<<" is pushed"<<endl;
cout<<str1[1]<<" is pushed"<<endl;
cout<<str1[2]<<" is pushed"<<endl;
cout<<str1[3]<<" is pushed"<<endl;
cout<<str1[4]<<" is pushed"<<endl;
cout<<"size of stack is:";
cout<<s.getTop()+1<<endl;
}
